### ISSUE TITLE
EnclaveCreatorHW::create_enclave generates duplicate enclave IDs

### ISSUE THREAD

**Issue by sean-jc** [01org/linux-sgx#67]
The logic in `EnclaveCreatorHW::create_enclave` that is responsible for generating unique IDs has multiple bugs.   The primary bug is that incrementing g_eid_low and consuming the result is not an atomic operation.  If two threads increment g_eid_low simultaneously, then both threads will see the same value of g_eid_low when generating the enclave_id.  This results in `CEnclavePool::add_enclave` failing because it detects duplicate IDs.  

The secondary bug is that the code is fundamentally flawed.  Splicing together two separate atomics is in itself not an atomic operation, i.e. there is no way to prevent race conditions where a thread sees the "wrong" g_eid_high value.

The immediate bug can be fixed by consuming `tmp` instead of `g_eid_low` when writing enclave_id.  Properly fixing all bugs requires adding/using a 64-bit atomic or wrapping the entire thing in a mutex.

```
  init: g_eid_low <= 1, g_eid_high <= 0
    T1: se_atomic_inc(&g_eid_low),  g_eid_low <= 2
    T2: se_atomic_inc(&g_eid_low),  g_eid_low <= 3
    T1: *enclave_id = ((uint64_t)g_eid_high << 32) | g_eid_low;  *enclave_id <= 3
    T2: *enclave_id = ((uint64_t)g_eid_high << 32) | g_eid_low;  *enclave_id <= 3
```

**enclave_creator_hw.cpp**
```C
    uint32_t tmp = se_atomic_inc(&g_eid_low);
    //32bit overflow
    if(0 == tmp)
        g_eid_high++;
    *enclave_id = ((uint64_t)g_eid_high << 32) | g_eid_low;
```

**Comment by lzha101** [comment]
Thanks for reporting this issue.  Will fix it in the next release.

**Comment by lzha101** [comment]
The bug is fixed in the latest 1.8 Release. Close this issue.
